lib/babe/build.go

```diff
@@ -270,6 +265,21 @@ func (b *BlockBuilder) buildBlockExtrinsics(slot Slot) []*transaction.ValidTrans
 			if _, ok := err.(*DispatchOutcomeError); !ok {
 				continue
 			}
+
+			// don't drop transactions that may be valid in a later block ie.
+			// run out of gas for this block or have a nonce that may be valid in a later block
+			var e *TransactionValidityError
+			if !errors.As(err, &e) {
+				continue
+			}
+
+			err = err.(*TransactionValidityError).msg
```

Comment:
This cast is not necessary since we already change the `err` into `TransactionValidityError` on line `273`

Reply:
it doesn't, the above lines do this:
```
			var e *TransactionValidityError
			if !errors.As(err, &e) {
				continue
			}
```

that doesn't cast it as a `*TransactionValidityError` afaik

Reply:
Sorry, the `errors.As` doesn't cast the variable `err`,  it gets the error and transform it to type `*TransactionValidityError` and sets to variable `&e`, without change the `err` variable, and returns true (or false if fails the type doesn't match). So basically it is possible to use `e.msg` without the cast on line `276`

from: https://golang.org/pkg/errors/#As
> ### func As(err error, target interface{}) bool
> **errors.As** finds the first error in err's chain that matches target, and if so, sets target to that error value and returns true. Otherwise, it returns false.

Reply:
okay makes sense, I updated it!